### PR TITLE
add benchmarks for prefix tree & run benches on travis

### PR DESCRIPTION
**Pull request by snd**
inspired by issue https://github.com/paritytech/paritydb/issues/22 i briefly looked for alternative counting bloom filters (or similar) that would truly support remove.

one promising is: https://github.com/seiflotfy/rust-cuckoofilter

it is quite a bit slower than @tomusdrw s prefix_tree however.

added benchmarks for cuckoofilter for comparison.

results on my machine:

```
test cuckoofilter_get_not_inserted_in_1          ... bench:          93 ns/iter (+/- 28)
test cuckoofilter_get_not_inserted_in_10000      ... bench:          94 ns/iter (+/- 29)
test cuckoofilter_get_not_inserted_in_1000000    ... bench:          94 ns/iter (+/- 28)
test cuckoofilter_get_random_inserted_in_1       ... bench:         105 ns/iter (+/- 23)
test cuckoofilter_get_random_inserted_in_10000   ... bench:         114 ns/iter (+/- 58)
test cuckoofilter_get_random_inserted_in_1000000 ... bench:         139 ns/iter (+/- 59)
test prefix_tree_get_not_inserted_in_1           ... bench:           4 ns/iter (+/- 1)
test prefix_tree_get_not_inserted_in_10000       ... bench:           4 ns/iter (+/- 1)
test prefix_tree_get_not_inserted_in_1000000     ... bench:           4 ns/iter (+/- 1)
test prefix_tree_get_random_inserted_in_1        ... bench:          15 ns/iter (+/- 4)
test prefix_tree_get_random_inserted_in_10000    ... bench:          18 ns/iter (+/- 15)
test prefix_tree_get_random_inserted_in_1000000  ... bench:          18 ns/iter (+/- 17)
```

prefix tree is really fast!

not happy with having to make `prefix_tree` mod public for it to be usable from the `benches` crate.
any suggestions?

### CODE REVIEW

File: benchmarks/benches/bloom.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+extern crate rand;
+extern crate paritydb;
+extern crate cuckoofilter;
+extern crate tempdir;
+extern crate itertools;
+
+use test::Bencher;
+use rand::{Rand, StdRng, SeedableRng};
+use rand::distributions::{IndependentSample, Range};
+use paritydb::prefix_tree::PrefixTree;
```
Comment (debris) [inline benchmarks/benches/bloom.rs]: I'd rather have `PrefixTree` public instead of making the whole module public

Comment (snd) [reply benchmarks/benches/bloom.rs]: done

**Comment by tomusdrw** [comment]
Note it's not entirely fair comparison :) Prefix tree with `key_index_bits=32` is probably 1GB in memory, while I'm sure that cuckoofilter is way smaller.

You can make it public and use `#[doc(hidden)]`? Or perhaps while compiling benches dependency a `#[cfg(test)]` is set?

**Comment by tomusdrw** [comment]
And BTW `prefix_tree` itself supports removal of entries, the problem is that the database needs to figure out if there is another key with the exact same prefix and remove only if there is none.

Alternatively we could count prefixes in the tree itself, but it will require way more memory (currently it's a simple bit vector)

**Comment by snd** [comment]
@tomusdrw it was not indented as a fair comparison :)
i had looked at #22 and seen that we'd get lots of false positives after many deletes.
so i looked at alternatives that "truly support" delete in the sense that they keep count.
the cuckoofilter is such a counting filter.
if the cuckoofilter had been close to the prefix tree in performance i would have proposed switching to it since it's counting and uses less space.
however it's quite a bit slower and i don't think we want to make that tradeoff.

think i'll document my thinking in the bench so others don't think it's intended as a direct comparison

**Comment by snd** [comment]
addressed the grumbles

out for review again

benchmarks are now run for the nightly tests as well as compared with benchmarking master (to check for regressions). you can see the results here: https://travis-ci.org/paritytech/paritydb/jobs/309652362#L1156

it would be nice to have a way to get notified on regressions on PRs without having to look into the travis build log. future work

**Comment by snd** [comment]
am i good to merge this?

**Pull request merged**